modules/core/src/main/java/org/apache/ignite/internal/util/distributed/SingleNodeMessage.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.util.distributed;
+
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.util.UUID;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.plugin.extensions.communication.Message;
+import org.apache.ignite.plugin.extensions.communication.MessageReader;
+import org.apache.ignite.plugin.extensions.communication.MessageWriter;
+
+/**
+ * Single node result message.
+ *
+ * @param <R> Result type.
+ */
+public class SingleNodeMessage<R extends Serializable> implements Message {
+    /** Serial version uid. */
+    private static final long serialVersionUID = 0L;
+
+    /** Initial channel message type (value is {@code 176}). */
+    public static final short TYPE_CODE = 176;
+
+    /** Process id. */
+    private UUID processId;
+
+    /** Process type. */
+    private int type;
+
+    /** Single node response. */
+    private R resp;
+
+    /** Error. */
+    private Exception err;
+
+    /** Empty constructor for marshalling purposes. */
+    public SingleNodeMessage() {
+    }
+
+    /**
+     * @param processId Process id.
+     * @param type Process type.
+     * @param resp Single node response.
+     * @param err Error.
+     */
+    public SingleNodeMessage(UUID processId, int type, R resp, Exception err) {
```

Comment:
It seems we should change type from int to the DistributedProcessType and use type.ordinal() if required.